tensorflow/contrib/tensorrt/kernels/trt_calib_op.cc

```diff
@@ -0,0 +1,167 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/contrib/tensorrt/kernels/trt_calib_op.h"
+#include "tensorflow/contrib/tensorrt/resources/trt_int8_calibrator.h"
+#include "tensorflow/contrib/tensorrt/resources/trt_resource_manager.h"
+#include "tensorflow/contrib/tensorrt/resources/trt_resources.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/framework/types.h"
+
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+#include "cuda_runtime_api.h"
+#include "tensorrt/include/NvInfer.h"
+
+namespace tensorflow {
+namespace trt {
+TRTCalibOp::TRTCalibOp(OpKernelConstruction* context) : OpKernel(context) {
+  OP_REQUIRES_OK(context, context->GetAttr("segment_nodes", &segment_nodes_));
+  OP_REQUIRES_OK(context, context->GetAttr("input_names", &input_names_));
+  OP_REQUIRES_OK(context, context->GetAttr("resource_name", &resource_name_));
+};
+
+void TRTCalibOp::Compute(tensorflow::OpKernelContext* ctx) {
+  auto trt_rm = tensorflow::trt::TRTResourceManager::instance();
+  auto res_mgr = trt_rm->getManager("TRTCalibOps");
+  tensorflow::trt::TRTCalibrationResource* calib_res = nullptr;
+  auto status = res_mgr->Lookup(resource_name_, resource_name_, &calib_res);
+
+  if (!status.ok()) {
+    ctx->SetStatus(status);
+    return;
+  }
+  int num_inputs = ctx->num_inputs();
+  // first run instantiate calibrator
+  if (calib_res->calibrator_ == nullptr) {
+    dev_tensors_.resize(num_inputs);
+    int batch_size = ctx->input(0).dim_size(0);
+    VLOG(1) << " Constructing calibrator";
+    for (int i = 0; i < num_inputs; i++) {
+      // allocate workspace on device for inputs
+      const tensorflow::Tensor& t = ctx->input(i);
+      OP_REQUIRES_OK(ctx,
+                     ctx->allocate_persistent(t.dtype(), t.shape(),
+                                              &dev_tensors_.at(i), nullptr));
+      const auto device_tensor = dev_tensors_.at(i).AccessTensor(ctx);
+      CHECK_EQ(t.TotalBytes(), device_tensor->TotalBytes());
+      void* device_address = nullptr;
+      {
+        auto tensor_type = device_tensor->dtype();
+        switch (tensor_type) {
+          case tensorflow::DT_FLOAT: {
+            device_address = (void*)device_tensor
+                                 ->flat<tensorflow::EnumToDataType<
+                                     tensorflow::DT_FLOAT>::Type>()
+                                 .data();
```

Comment:
Hi @samikama, I think we need a 'break' here and in all other cases? Also this file cannot build now. Since I'm trying to fix the build internally and I already fix this part, I'll push it out. So please fetch the head and let me know if it doesn't work on your machine.

Reply:
That is correct sorry I missed that. The reason I chose macro was due to the discussion about the making copies of weights in previous pr. My understanding was that the tensors are not necessarily contiguous and thus we needed to make a copy of it. Thus I assumed flat() would need to make a copy in order for it to be contiguous. If we pass get the flat in another function, we have to ensure the lifetime of the Flat object. Am I wrong? Are tensors contiguous and Flat objects are just wrappers around internal buffer? If not, how are you ensuring that the memory is still there when you return from the function?

Reply:
Actually Tensor are contiguous, and flat() is used to return the pointer to the underlying buffer. What I was saying is that [TensorProto::tensor_content](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.proto#L36) is not contiguous internally (although it is externally) so we need to copy it out. Sorry for the confusion.